modules/tftp/server.rb

```diff
@@ -90,21 +103,107 @@ def pxe_default
     end
 
     def pxeconfig_file(mac)
-      ["#{pxeconfig_dir}/menu.lst.01" + mac.delete(':').upcase, "#{pxeconfig_dir}/01-" + mac.tr(':', '-').upcase]
+      ["#{pxeconfig_dir}/menu.lst.01" + mac.delete(':').upcase, "#{pxeconfig_dir}/01-" + dashed_mac(mac).upcase]
     end
   end
 
   class Pxegrub2 < Server
-    def pxeconfig_dir
-      "#{path}/grub2"
+    def bootloader_path(os, release, arch)
+      bootloader_path = File.join(Proxy::TFTP::Plugin.settings.tftproot, 'bootloader-universe/pxegrub2', os, release, arch)
+
+      logger.debug "TFTP: Checking if bootloader universe is configured for for OS '#{os}' version '#{release}' (#{arch})."
+      logger.debug "TFTP: Checking if directory '#{bootloader_path}' exists."
+      unless Dir.exist?(bootloader_path)
+        logger.debug "TFTP: Directory '#{bootloader_path}' does not exist."
+
+        bootloader_path = File.join(Proxy::TFTP::Plugin.settings.tftproot, 'bootloader-universe/pxegrub2', os, 'default', arch)
+        logger.debug "TFTP: Checking if directory '#{bootloader_path}' exists."
+        unless Dir.exist?(bootloader_path)
+          logger.debug "TFTP: Directory '#{bootloader_path}' does not exist."
+          return
+        end
+      end
+
+      bootloader_path
+    end
+
+    # Configures bootloader files for a host in its host-config directory
+    #
+    # @param mac [String] The MAC address of the host
+    # @param os [String] The lowercase name of the operating system of the host
+    # @param release [String] The major and minor version of the operating system of the host
+    # @param arch [String] The architecture of the operating system of the host
+    # @param bootfilename_efi [String] The architecture specific boot filename suffix
+    def setup_bootloader(mac:, os:, release:, arch:, bootfilename_efi:)
+      pxeconfig_dir_mac = pxeconfig_dir(mac)
+
+      logger.debug "TFTP: Deploying host specific bootloader files to '#{pxeconfig_dir_mac}'."
+
+      FileUtils.mkdir_p(pxeconfig_dir_mac)
+      FileUtils.rm_f(Dir.glob("#{pxeconfig_dir_mac}/*.efi"))
+
+      bootloader_path = bootloader_path(os, release, arch)
+
+      if bootloader_path
+        logger.debug "TFTP: Creating symlinks from bootloader universe."
+
+        Dir.glob(File.join(bootloader_path, '*.efi')).each do |source_file|
+          file_name = File.basename(source_file)
+          symlink_path = File.join(pxeconfig_dir_mac, file_name)
+
+          logger.debug "TFTP: Creating relative symlink: '#{symlink_path}' -> '#{source_file}'"
+          FileUtils.ln_sr(source_file, symlink_path, force: true)
```

Comment:
I'm using Ruby version 3.0.4, and I don't have access to `FileUtils.ln_sr`. This method has been available since FileUtils v1.7.0: [FileUtils v1.7.0 release](https://github.com/ruby/fileutils/releases/tag/v1.7.0), but I currently have v1.6.0

Reply:
We should support FileUtils version 1.4.1 and above, as discussed [here](https://matrix.to/#/!dxfWqeLNsImROheGXo:matrix.org/$1723561908682oUJpM:matrix.org?via=matrix.org&via=ansible.im&via=ctrl-c.liu.se).

Reply:
Hey Nofar, I now headed for a new approach for the relative symlinks. I calculate the relative paths with the `Pathname.relative_path_from` method. According to the [Ruby API documentation](https://rubyapi.org/2.7/o/pathname#method-i-relative_path_from) it is available in Ruby 2.7 and newer.

Note: I can't remember if I mentioned it, but I saw that the suggested `File.relative_path` method (DM) is provided by the `yard` RubyGem. As far as I saw, it is not a dependency of Smart Proxy and also it was not installed on my Test Deployment.